File: revert-list.php
<?php

    /**
     * Nao tem como criar classes internas no PHP,
     * entao infelizmente a classe Node fica publica
     * ¯\_(ツ)_/¯
     */
    class Node {
        /**
         * @var $next Node
         * @var $content ?
         */
        public $content;
        public $next;

        function __construct($content) {
            $this->content = $content;
            $this->next = null;
        }
    }

    class LinkedList {
        /**
         * @var $count int
         * @var $head Node
         */
        private $count;
        private $head;

        function __construct() {
            $this->count = 0;
            $this->head = null;
        }

        /**
         * @return void
         */
        function add($item) {
            /**
             * @var $node Node
             */
            $node = new Node($item);
            if ($this->head == null)
                $this->head = $node;
            else {
                /**
                 * @var $n Node
                 */
                $n = $this->head;
                while ($n->next != null)
                    $n = $n->next;

                $n->next = $node;
            }

            $this->count++;
        }

        /**
         * @return void
         */
        function revert() {
            /**
             * @var $prev Node
             * @var $curr Node
             */
            $prev = null;
            $curr = $this->head;

            while ($curr != null) {
                $tmp = $curr->next;
                $curr->next = $prev;
                $prev = $curr;
                $curr = $tmp;
            }

            $this->head = $prev;
        }

        function item($index) {

            /**
             * @var $n Node
             */
            $n = $this->head;

            for ($i = 0; $i < $index; $i++)
                $n = $n->next;

            return $n->content;
        }

        function size() {
            return $this->count;
        }

        function __toString() {
            $buf = "";
            for ($i = 0; $i < $this->size(); $i++)
                $buf .= $this->item($i) . " ";

            return $buf . PHP_EOL;
        }
    }

    $list = new LinkedList();
    for ($i = 0; $i < 10; $i++)
        $list->add($i + 1);

    /**
     * print eh uma funcao nativa do PHP, (mesma coisa que echo)
     * entao implementamos o metodo __toString para formatar a saida
     */

    print($list);
    $list->revert();
    print($list);

    $str_list = new LinkedList();
    for ($i = 0; $i < 5; $i++)
        $str_list->add("str" . ($i + 1));

    print($str_list);
    $str_list->revert();
    print($str_list);
    
    echo PHP_EOL;
    

File: revert-list.cpp
#include <iostream>
#include <sstream>

using namespace std;

template<class T>
class LinkedList {
	private:
		class Node {
			public:
				T content;
				Node *next;
				Node(T content);
		};
	
		int count;
		Node *head;
	public:
		LinkedList() {
			count = 0;
			head = NULL;
		}
	
		void add(T item);
		void revert();
		T item(int index);
		int size();
};

template<class T>
LinkedList<T>::Node::Node(T content) {
	this->content = content;
	next = NULL;
}

template<class T>
void LinkedList<T>::add(T item) {
	Node *node = new Node(item);
	if(head == NULL)
		head = node;
	else {
		Node *n = head;
		while(n->next!=NULL)
			n = n->next;
		
		n->next = node;
	}
	
	count++;
}

template<class T>
void LinkedList<T>::revert() {
	Node *prev = NULL, *curr = head;
	while(curr!=NULL) {
		Node *tmp = curr->next;
		curr->next = prev;
		prev = curr;
		curr = tmp;
	}
	
	head = prev;
}

template<class T>
T LinkedList<T>::item(int index) {

	Node *n = head;

	for(int i=0; i<index; i++)
		n = n->next;
		
	return n->content;
}

template<class T>
int LinkedList<T>::size() {
	return count;
}

template<class T>
string print(LinkedList<T> &list) {
	stringstream buf;

	for(int i=0; i<list.size(); i++)
		buf << list.item(i) << " ";
		
	buf << endl;
	
	return buf.str();
}

int main() {
	
	LinkedList<int> list;
	for(int i=0; i<10; i++)
		list.add(i+1);
		
	cout << print(list);
	list.revert();
	cout << print(list);
		
	LinkedList<string> str_list;
	for(int i=0; i<5; i++)
		str_list.add("str" + to_string(i+1));
		
	cout << print(str_list);
	str_list.revert();
	cout << print(str_list);
	
	cout << endl;
}
